torch/distributed/algorithms/join.py

```diff
@@ -213,3 +270,55 @@ def _notify_procs_to_terminate(self):
         # and throws a `RuntimeError` in Python 3.7+ (PEP 479), so we just
         # raise a `RuntimeError` here
         raise RuntimeError(f"Rank {self._rank} exhausted all inputs.")
+
+    @staticmethod
+    def notify_join_context(joinable: _Joinable):
+        r"""
+        Notifies the join context manager that the calling process has not yet
+        joined; then, if ``throw_on_early_termination=True``, checks if uneven
+        inputs have been detected (i.e. if one process has already joined) and
+        throws an exception if so.
+
+        This method should be called from a :class:`_Joinable` object before
+        its per-iteration collective communications. For example, this should
+        be called at the beginning of the forward pass in DDP.
+
+        Only the first :class:`_Joinable` object passed into the context
+        manager performs the collective communications in this method, and
+        for the others, this method is vacuous.
+
+        Arguments:
+            joinable (_Joinable): the :class:`_Joinable` object calling this
+                method.
+
+        Returns:
+            An async work handle for the all-reduce meant to notify the context
+            manager that the process has not yet joined if ``joinable`` is the
+            first one passed into the context manager; ``None`` otherwise.
+        """
+        if not hasattr(joinable, "_join_config"):
```

Comment:
Perhaps, this should be changed to an `assert` since if `hasattr(joinable, "_join_config") == False`, then most likely the `_Joinable` constructor was not called, which should not be silent.

Reply:
Changed to an `assert` locally; will be visible in the next push.